File: DFS/079.Word-Search/079.Word-Search.ts
function exist(board: string[][], word: string): boolean {
    const m = board.length;
    const n = board[0].length;
    const used = new Array(m).fill(null).map(() => new Array(n).fill(false));

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (dfs(i, j, 0)) {
                return true;
            }
        }
    }
    return false;

    function dfs(i: number, j: number, wordCursor: number) {
        if (i < 0 || i >= m || j < 0 || j > n) {
            return false;
        }

        if (used[i][j] || word[wordCursor] !== board[i][j]) {
            return false;
        }

        if (wordCursor == word.length - 1) {
            return true
        }

        used[i][j] = true

        const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

        for (const [row, column] of directions) {
            if (dfs(i + row, j + column, wordCursor + 1)) {
                return true
            }
        }

        used[i][j] = false
    }
};

File: DFS/079.Word-Search/079.Word-Search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])

        seen = [[0] * n for i in range(m)]

        def dfs(row, column, index):
            if row < 0 or row >= m or column < 0 or column >= n:
                return False

            if seen[row][column] or board[row][column] != word[index]:
                return False

            if index == len(word) - 1:
                return True

            seen[row][column] = 1

            directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            for [x, y] in directions:
                if dfs(row + x, column + y, index + 1):
                    return True

            seen[row][column] = 0
            return False

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False
            
